Initial commit for Iceberg catalog migrator

ajantha-bhat opened the pull request:
Please feel free to checkout this PR and run locally. 
Verified `gradle clean build` and all the tests are passing. 

Changes: 
- Copied the latest `main` branch code form https://github.com/projectnessie/iceberg-catalog-migrator for donation
- Replace headers to use apache headers
- rename packages to `org.apache.polaris.iceberg.catalog.migrator`
- update security.md
- remove the automated workflow related to projectnessie accounts
- clean up dremio keyword
- Adopt to polaris tools repo

TODO: (In the follow up PRs)

 - Once Polaris release available, add tests for polaris
 - Update readme to use polaris examples
 - Finalize release version
 - Add automated release workflow
 - Double check LICENSE and NOTICE (add a separate one for CLI runtime jar distribution)
 - Bump versions (like Iceberg, Nessie etc)
 - Migrate open issues/proposals from original repo
 - Build enhancements (like dependency between javadoc and jandex tasks)
 
 This PR also contains the contributions from

Co-authored-by: Robert Stupp <snazy@snazy.de>
Co-authored-by: Christopher Lambert <xn137@gmx.de>
Co-authored-by: Harsh Maheshwari <57615655+harshm-dev@users.noreply.github.com>

ajantha-bhat commented:
cc: @jbonofre, @snazy, @dimas-b

mansehajsingh commented:
Should there be a `gradle-wrapper.jar` in `gradle/wrapper/` trying to run a gradle build fails unlike in the original `iceberg-catalog-migrator` project, where `gradle/wrapper/gradle-wrapper.jar` exists.

mansehajsingh commented on .gitignore:
```diff
@@ -0,0 +1,74 @@
+# gradle
+.gradle/
+build/
+gradle/wrapper/gradle-wrapper.jar
```
It may have failed to copy over `gradle-wrapper.jar` since it has been ignored here

ajantha-bhat replied:
forcefully added the jar now

mansehajsingh commented on settings.gradle.kts:
```diff
@@ -0,0 +1,49 @@
+
+fun catalogMigratorProject(name: String) {
+  include("iceberg-catalog-migrator-$name")
+  project(":iceberg-catalog-migrator-$name").projectDir = file(name)
```
Should this be `file(iceberg-catalog-migrator/$name)` now that `iceberg-catalog-migrator-api`, `iceberg-catalog-migrator-api-test`, `iceberg-catalog-migrator-cli` are in the `iceberg-catalog-migrator` subfolder relative to the `settings.gradle`?

sfc-gh-mansingh commented on iceberg-catalog-migrator/cli/build.gradle.kts:
```diff
@@ -0,0 +1,155 @@
+    )
+  }
+
+val mainClassName = "org.apache.polaris.catalogs.migrator.cli.CatalogMigrationCLI"
```
is this still the correct package name?

mansehajsingh replied:
is this still the right package name?

ajantha-bhat commented:
@mansehajsingh: Thanks for taking a look. I have addressed the comments.

snazy reviewed (changes requested):
Overall the PR LGTM.
My request for changes is due to the addition of a .jar file and mention of the donation.

I also think, that the migrator should actually live in its own directory, as this repository is rather generic ("tools"). I propose to either move everything to `iceberg-catalog-migrator/` or to have a dedicated Git repo for it.

ajantha-bhat commented on NOTICE:
```diff
@@ -0,0 +1,10 @@
+The Apache Software Foundation (http://www.apache.org/).
+
+The initial code for the Polaris project was donated
+to the ASF by Snowflake Inc. (https://www.snowflake.com/) copyright 2024.
```
good catch. Updated.

ajantha-bhat commented:
@snazy: thanks for the review. I have updated the gradle stuff to be similar to polaris main repo and verified `./gradlew clean build` and handled other comments.

ajantha-bhat commented:
And regarding this

> I also think, that the migrator should actually live in its own directory, as this repository is rather generic ("tools"). I propose to either move everything to iceberg-catalog-migrator/ or to have a dedicated Git repo for it.

The modules related to catalog migrator is under `iceberg-catalog-migrator/` already. Do you mean move other folders like bom, buildsrc, codestyle also to `iceberg-catalog-migrator/`? I think these can be common for other tools that will be hosted in this repository.

jbonofre reviewed (commented):
First quick pass, overall ok, a couple of nits.

I'm also checking build and generated artifacts (in terms of legal).

dimas-b commented on gradle/libs.versions.toml:
```diff
@@ -0,0 +1,64 @@
+
+[libraries]
+assertj = { module = "org.assertj:assertj-core", version.ref = "assertj" }
+aws-sdk-apache-client = { module = "software.amazon.awssdk:apache-client", version.ref = "aws" }
```
We should probably mention these in LICENSE, WDYT?

ajantha-bhat replied:
> runtimeOnly(libs.aws.sdk.apache.client)

So, user has to supplement these jars. We are not shipping it. 

That being said. We do need to have LICENSE and NOTICE for the runtime jar (CLI) that we are shipping. Maybe in a follow up PR.

dimas-b replied:
Handling binary LICENSE later SGTM :+1:

ajantha-bhat replied:
added in the PR description of TODO to track it.

ajantha-bhat commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,3 @@
+This is a project of the [Apache Software Foundation](https://apache.org/) and follows the [ASF Code of Conduct](https://www.apache.org/foundation/policies/conduct).
```
ok. removed

ajantha-bhat commented on LICENSE:
```diff
@@ -0,0 +1,202 @@
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
```
added about gradle wrapper

jbonofre commented on gradle/contributors.csv:
If you mean adding these contributors as `collaborators` in `.asf.yaml`, yes.
if you mean adding these contributors as committers, it has to be approved by the PPMC.

Usually, we invite as committer the author of the donation PR, and we add later (by meritocracy).

dimas-b replied:
collaborators SGTM.

... but why do we have this `contributors.csv` file?

ajantha-bhat replied:
removed `developers.cvs` and `contributors.csv`. 

Maybe I can copy the initial folder with git history to retain other contribution history?

ajantha-bhat replied:
In PR description I have added now

This PR also contains the contributions from

```
Co-authored-by: Robert Stupp [snazy@snazy.de](mailto:snazy@snazy.de)
Co-authored-by: Christopher Lambert [xn137@gmx.de](mailto:xn137@gmx.de)
Co-authored-by: Harsh Maheshwari [57615655+harshm-dev@users.noreply.github.com](mailto:57615655+harshm-dev@users.noreply.github.com)
```

Looks like the email that Harsh has used (dremio email) is no longer linked to github. 

Once, I have approval for this PR. I will squash and update the commit message about coauthor.

dimas-b replied:
Thx! LGTM!

jbonofre commented on buildSrc/build.gradle.kts:
```diff
@@ -0,0 +1,46 @@
+
+dependencies {
+  implementation(gradleKotlinDsl())
+  implementation(baselibs.spotless)
+  implementation(baselibs.jandex)
+  implementation(baselibs.idea.ext)
+  implementation(baselibs.shadow)
+  implementation(baselibs.errorprone)
```
`LICENSE` in the root folder is for source distribution. Here's these dependencies are at build time and not bundled in the our distributed artifact.
So, no need to mention in `LICENSE`.

mansehajsingh commented:
> And regarding this
> 
> > I also think, that the migrator should actually live in its own directory, as this repository is rather generic ("tools"). I propose to either move everything to iceberg-catalog-migrator/ or to have a dedicated Git repo for it.
> 
> The modules related to catalog migrator is under `iceberg-catalog-migrator/` already. Do you mean move other folders like bom, buildsrc, codestyle also to `iceberg-catalog-migrator/`? I think these can be common for other tools that will be hosted in this repository.

+1 to this - would be super helpful if other additions to the repo were able to share these and deduplicate the work needed to bootstrap a new tool.

ajantha-bhat commented:
Hi everyone, Thanks for the reviews. 

I have addressed all the comments. 
As per TODO listed in PR description, we can have follow up PRs. 

I would like to squash the PR once I have approval to add `coauthored by` to the commit message. 
Please check.

ajantha-bhat commented:
Also merging this can help for #2 to avoid duplicate work of introducing build and infra related stuffs.

mansehajsingh reviewed (approved):
Accidentally submitted approval on my other account.

ajantha-bhat commented:
I squashed the commit to add co-author by in the commit message. 
( I have a backup. If people want to still review it individually again).

ajantha-bhat commented:
@snazy: Could you please take another look?

jbonofre commented on README.md:
```diff
@@ -0,0 +1,5 @@
+This repo has the tools developed by [Apache polaris community](https://polaris.apache.org/)
```
Can you please add ASF header here ?

ajantha-bhat replied:
Done. Added for both the README in the repo.

snazy commented on gradlew.bat:
```diff
@@ -0,0 +1,94 @@
+@rem
```
Please remove this file

snazy commented on .github/workflows/main.yml:
Better rename to `.github/workflows/iceberg-catalog-migrator-main.yml`

snazy commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,61 @@
+# under the License.
+#
+
+name: CI
```
```suggestion
name: Iceberg Catalog Migrator CI
```

snazy commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,61 @@
+
+on:
+  push:
+    branches: [ main ]
```
```suggestion
    branches: [ main ]
    paths:
      - `iceberg-catalog-migrator/**'
```

snazy commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,61 @@
+on:
+  push:
+    branches: [ main ]
+  pull_request:
```
```suggestion
  pull_request:
    paths:
      - `iceberg-catalog-migrator/**'
```

snazy commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,61 @@
+        uses: gradle/actions/setup-gradle@v4
+
+      - name: Build & Check
+        run: ./gradlew --rerun-tasks assemble ${{ env.ADDITIONAL_GRADLE_OPTS }} check publishToMavenLocal
```
```suggestion
        run: |
          cd iceberg-catalog-migrator
          ./gradlew --rerun-tasks assemble ${{ env.ADDITIONAL_GRADLE_OPTS }} check publishToMavenLocal
```

snazy commented on ide-name.txt:
```diff
@@ -0,0 +1 @@
+polaris-tools
```
```suggestion
Catalog Migrator
```

snazy reviewed (commented):
Given that the polaris-tools repository is meant to host multiple and independent projects, I think that all those sub-projects should live in their own subfolder and be completely independent.

Therefore, everything, except the `.github/` folder and the `README.md` + `SECURITY.md` files should go under the `iceberg-catalog-migrator/` folder.

snazy commented on iceberg-catalog-migrator/settings.gradle.kts:
```diff
@@ -0,0 +1,47 @@
+
+val baseVersion = file("version.txt").readText().trim()
+
+rootProject.name = "polaris-tools"
```
```suggestion
rootProject.name = "polaris-iceberg-catalog-migrator"
```

ajantha-bhat replied:
done

snazy reviewed (commented):
Hm, the build's failing:

```
FAILURE: Build failed with an exception.

* Where:
Settings file '/home/snazy/devel/polaris/polaris-tools/iceberg-catalog-migrator/settings.gradle.kts' line: 20

* What went wrong:
/home/snazy/devel/polaris/polaris-tools/iceberg-catalog-migrator/version.txt (No such file or directory)
```

That and a last name change and we should be good!

Don't forget to open an issue about add support for releases.

snazy commented:
Issue #5 created

ajantha-bhat commented:
> Hm, the build's failing:

I verified locally before pushing that builds are passing. But due to `.gitignore` file, `version.txt `was not committed to the repo. Forcefully committed it and verified the build by checking out to a fresh repo. 

I think PR is ready.

snazy reviewed (approved):
LGTM!
